streaming/src/main/scala/org/apache/spark/streaming/scheduler/StreamingListenerBus.scala

```diff
@@ -17,20 +17,28 @@
 
 package org.apache.spark.streaming.scheduler
 
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.{LinkedBlockingQueue, CopyOnWriteArrayList}
+
+import scala.util.control.NonFatal
+
 import org.apache.spark.Logging
-import scala.collection.mutable.{SynchronizedBuffer, ArrayBuffer}
-import java.util.concurrent.LinkedBlockingQueue
+import org.apache.spark.util.Utils
 
 /** Asynchronously passes StreamingListenerEvents to registered StreamingListeners. */
 private[spark] class StreamingListenerBus() extends Logging {
-  private val listeners = new ArrayBuffer[StreamingListener]()
-    with SynchronizedBuffer[StreamingListener]
+  // `listeners` will be set up during the initialization of the whole system and the number of
+  // listeners is small, so the copying cost of CopyOnWriteArrayList will be little. With the help
+  // of CopyOnWriteArrayList, we can eliminate a lock during processing every event comparing to
+  // SynchronizedBuffer.
+  private val listeners = new CopyOnWriteArrayList[StreamingListener]()
```

Comment:
is the use of copy on write here to speed up initialization, in the sense that events posted on start up will not be blocked as we add listeners? I may be missing something, but this speed up doesn't seem very significant to me.